clang/test/lab1/loginov_maxim/test.cpp

```diff
@@ -0,0 +1,176 @@
+// NON_EXIST_FUNC: void func2() {
+// NON_EXIST_FUNC-NEXT: int c = func(2);
+// NON_EXIST_FUNC-NEXT: int b = func(c) + func(3);
+// NON_EXIST_FUNC-NEXT: }
+
+// RUN: %clang_cc1 -load %llvmshlibdir/RenameIdentificatorPlugin%pluginext\
+// RUN: -add-plugin Rename\
+// RUN: -plugin-arg-Rename type=class\
+// RUN: -plugin-arg-Rename cur-name=Base\
+// RUN: -plugin-arg-Rename new-name=SimpleClass %t/rename_class.cpp
+// RUN: FileCheck %s < %t/rename_class.cpp --check-prefix=CLASS
+
+// CLASS: class SimpleClass{
+// CLASS-NEXT: private:
+// CLASS-NEXT: int a;
+// CLASS-NEXT: int b;
+// CLASS-NEXT: public:
+// CLASS-NEXT: SimpleClass() {}
+// CLASS-NEXT: SimpleClass(int a, int b): a(a), b(b) {}
+// CLASS-NEXT: ~SimpleClass();
+// CLASS-NEXT: };
+// CLASS: void func() {
+// CLASS-NEXT: SimpleClass a;
+// CLASS-NEXT: SimpleClass* var = new SimpleClass(1, 2);
+// CLASS-NEXT: delete var;
+// CLASS-NEXT: }
+
+// RUN: %clang_cc1 -load %llvmshlibdir/RenameIdentificatorPlugin%pluginext\
+// RUN: -add-plugin Rename\
+// RUN: -plugin-arg-Rename help\
+// RUN: 2>&1 | FileCheck %s --check-prefix=HELP
+
+// HELP: Specify three required arguments:
+// HELP-NEXT: -plugin-arg-Rename type=["var", "func", "class"]
+// HELP-NEXT: -plugin-arg-Rename cur-name="Current identifier name"
+// HELP-NEXT: -plugin-arg-Rename new-name="New identifier name"
+
+// RUN: not %clang_cc1 -load %llvmshlibdir/RenameIdentificatorPlugin%pluginext\
+// RUN: -add-plugin Rename\
+// RUN: -plugin-arg-Rename cur-name=B\
+// RUN: -plugin-arg-Rename new-name=C\
+// RUN: 2>&1 | FileCheck %s --check-prefix=ERROR
+
+// RUN: not %clang_cc1 -load %llvmshlibdir/RenameIdentificatorPlugin%pluginext\
+// RUN: -add-plugin Rename\
+// RUN: -plugin-arg-Rename cur-name=B\
+// RUN: -plugin-arg-Rename new-name=C\
+// RUN: -plugin-arg-Rename param=val\
+// RUN: 2>&1 | FileCheck %s --check-prefix=ERROR
+
+// RUN: not %clang_cc1 -load %llvmshlibdir/RenameIdentificatorPlugin%pluginext\
+// RUN: -add-plugin Rename\
+// RUN: -plugin-arg-Rename type=undefined\
+// RUN: -plugin-arg-Rename cur-name=B\
+// RUN: -plugin-arg-Rename new-name=C\
+// RUN: 2>&1 | FileCheck %s --check-prefix=ERROR
+
+// RUN: not %clang_cc1 -load %llvmshlibdir/RenameIdentificatorPlugin%pluginext\
+// RUN: -add-plugin Rename\
+// RUN: 2>&1 | FileCheck %s --check-prefix=ERROR
+
+//ERROR: Invalid arguments
+//ERROR-NEXT: Specify "-plugin-arg-Rename help" for usage
+
+//--- rename_var.cpp
+int func() {
+  int a = 2, b = 2;
+  a = b + a;
+  a++;
+  return a;
+}
+//--- rename_non_existent_var.cpp
+int func() {
+  int a = 2;
+  int b = 3;
+  b += a;
+  a++;
+  return b - a;
+}
+//--- rename_func.cpp
+int function(int param) {
+    int a;
+    a = 2;
+    return param + a;
+}
+int other_func(){
+  function(3);
+  int a = function(2) + 3;
+  return a;
+}
+//--- rename_non_existent_func.cpp
+int func(int a) {
+  int b = 2;
+  return a + b;
+}
+
+void func2() {
+  int c = func(2);
+  int b = func(c) + func(3);
+}
+//--- rename_class.cpp
+class Base{
```

Comment:
Could you please add a test of renaming template class name:
```suggestion
void func() {
  template<typename T>
  class TClass {
     T field;
  };

  TClass<int> tc;
}
```
Rename TClass into ClassT, for example.